src/data_generators/tests/edp_data_generator_test.py

```diff
@@ -0,0 +1,117 @@
+                "mean": [5.0],
+                "std": [10.0],
+                "cardinality": [3],
+            }
+        )
+
+        expected_data = pd.DataFrame(
+            {
+                "Publisher ID": [2] * 5,
+                "Event ID": [0, 1, 2, 3, 4],
+                "Sex": ["M"] * 5,
+                "Age Group": ["18-24"] * 5,
+                "Social Grade": ["ABC1"] * 5,
+                "Date": ["2022/01/02"] * 5,
+                "Complete": ["1"] * 5,
+                "VID": [0, 2, 2, 1, 0],
+            }
+        )
+        data = EdpDataGenerator(data_design, 1).generate_data()
+        pd.testing.assert_frame_equal(data, expected_data)
+
+    def test_single_publisher_multiple_demos(self):
+        data_design = pd.DataFrame(
+            {
+                "publisher_id": [2] * 3,
+                "mc_id": ["mc-5"] * 3,
+                "gender": ["M", "F", "M"],
+                "age_group": ["18-24", "25-34", "35-44"],
+                "social_grade": ["ABC1", "ABC2", "ABC3"],
+                "date": ["2022/01/02", "2022/01/03", "2022/01/03"],
+                "mean": [5.0] * 3,
+                "std": [10.0] * 3,
+                "cardinality": [2] * 3,
+            }
+        )
+
+        expected_data = pd.DataFrame(
+            {
+                "Publisher ID": [2] * 7,
+                "Event ID": [0, 1, 2, 3, 4, 5, 6],
+                "Sex": ["M", "M", "M", "F", "F", "M", "M"],
+                "Age Group": ["18-24"] * 3 + ["25-34"] * 2 + ["35-44"] * 2,
+                "Social Grade": ["ABC1"] * 3 + ["ABC2"] * 2 + ["ABC3"] * 2,
+                "Date": ["2022/01/02"] * 3 + ["2022/01/03"] * 4,
+                "Complete": ["1"] * 7,
+                "VID": list([0, 0, 1, 3, 2, 5, 4]),
+            }
+        )
+        data = EdpDataGenerator(data_design, 1).generate_data()
+        pd.testing.assert_frame_equal(data, expected_data)
+
+    def test_multiple_publishers_multiple_demos(self):
+        data_design = pd.DataFrame(
+            {
+                "publisher_id": [1, 2, 1],
+                "mc_id": ["mc-5"] * 3,
+                "gender": ["M", "M", "F"],
+                "age_group": ["18-24", "18-24", "25-34"],
+                "social_grade": ["ABC1", "ABC1", "ABC2"],
+                "date": ["2022/01/02", "2022/01/02", "2022/01/03"],
+                "mean": [5.0] * 3,
+                "std": [10.0] * 3,
+                "cardinality": [2] * 3,
+            }
+        )
+
+        expected_data = pd.DataFrame(
+            {
+                "Publisher ID": [1, 1, 2, 2, 2, 1, 1],
+                "Event ID": [0, 1, 2, 3, 4, 5, 6],
+                "Sex": ["M", "M", "M", "M", "M", "F", "F"],
+                "Age Group": ["18-24"] * 5 + ["25-34"] * 2,
+                "Social Grade": ["ABC1"] * 5 + ["ABC2"] * 2,
+                "Date": ["2022/01/02"] * 5 + ["2022/01/03"] * 2,
+                "Complete": ["1"] * 7,
+                "VID": list([0, 3, 1, 2, 3, 5, 4]),
```

Comment:
Could you help me understand this result?  Here pub 2 reached three vids 1, 2, 3.  But above, the cardinality was set as 2.

Reply:
In my view, this is actually a bug in GaussianCopulaDataSet.  By the way, to be consistent with IndependentOverlapDataSet and SequentiallyCorrelatedOverlapDataSet, this should be called GaussianCopulaOverlapDataSet, and it should inherit from OverlapDataSet.  In any event, it would be preferable if GaussianCopula(Overlap)DataSet guaranteed that the cardinality of each publisher in the output set is identical to the cardinality of the corresponding publisher in the input data set.  It would be even better if it could guarantee that the marginal distributions are the same, but I don't see an easy way to guarantee that.

Reply:
Thanks! Makes great sense.  Two initial responses:
1.  Good catch, the copula data set only has **approximate** single pub agreement, because the data set is obtained by random sampling.  Like you said, I didn't see an easy way either to guarantee exact agreement.
2. I've thought about making CopulaDataSet a subclass of OverlapDataSet and it didn't seem very easy to refactor `OverlapDataSet`.   I regret not having a great design of `OverlapDataSet` a year ago.
Anyway, the discrepancy here is well explained, so I'm approving your PR.  For the above 2 issues, LMK if any idea pops up in your mind.

Reply:
Thanks for reviewing and approving the code.  I think it would be easy to modify CopulaDataSet so that the single publisher data sets in the output have the same cardinality as in the input.  This could be done as follows:

1.  Generate enough random samples such that the total cardinality of each publisher in the output set equals or exceeds the total cardinality in the input set.  
2. For each publisher where the cardinality in the output set exceeds the cardinality in the input set, repeat the following:  Choose a user at random and delete all of the impressions of that user in the given publisher.

The joint distribution of the resulting data set will deviate very slightly from the true distribution defined by the Gaussian copula.

Reply:
Good idea:) Will modify PR 108 tomorrow

Reply:
From today's meeting I think I have some more urgent AIs, for example, enabling the check of single pub k+ agreement in experimental_trial.  So, I'll defer the improvement of `CopulaDataSet`, plmk if there's any objection.  (If by any chance you have bandwidth to realize this improvement, it'll be awesome!)